hw_lesson13.py

```diff
@@ -0,0 +1,107 @@
+import inspect
+from functools import wraps
+
+"""
+Task 1
+Write a decorator that prints a function with arguments passed to it.
+NOTE! It should print the function, not the result of its execution!
+"""
+
+
+def logger(func):
+    def wrap():
+        code_func = inspect.getsource(func)
+
+        split_for_return = code_func.split('return')
+
+        give_code = split_for_return[-1]
+
+        return give_code
+
+    return print(wrap())
+
+
+@logger
+def add(x, y):
+    return x + y
+
+
+@logger
+def square_all(*args):
+    return [arg ** 2 for arg in args]
+
+
+"""
+Task 2
+
+Write a decorator that takes a list of stop words and replaces them with * inside the decorated function
+"""
+
+
+def stop_words(words: list):
+    def wrap(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            string: str = func(*args, **kwargs)
+
+            for i in range(len(words)):
+                string = string.replace(words[i], '*')
+
+            return string
+
+        return wrapper
+
+    return wrap
+
+
+@stop_words(['pepsi', 'BMW'])
+def create_slogan(name: str) -> str:
+    return f"{name} drinks pepsi in his brand new BMW!"
+
+
+assert create_slogan("Steve") == "Steve drinks * in his brand new *!"
+
+
+"""
+Task 3
+
+Write a decorator `arg_rules` that validates arguments passed to the function.
+A decorator should take 3 arguments:
+
+max_length: 15
+
+type_: str
+
+contains: [] - list of symbols that an argument should contain
+
+If some of the rules' checks returns False, the function should return False and print the reason it failed;
+otherwise, return the result.
+"""
+
+
+def arg_rules(type_: type, max_length: int, contains: list):
+    def wrap_func(func):
+        @wrap_func(func)
+        def wrapper(*args, **kwargs):
+            name: str = func(*args, **kwargs)
+
+            new_str = set(''.join(contains))
+
+            if type(name) == type_ and len(name) < max_length and new_str.issubset(set(name)):
+                return True
+            else:
+                return False
+
+        return wrapper
+
+    return wrap_func
+
+
+@arg_rules(type_=str, max_length=15, contains=['05', '@'])
+def create_slogan(name: str) -> str:
+    return f"{name} drinks pepsi in his brand new BMW!"
+
+
+assert create_slogan('johndoe05@gmail.com') is False
+
+assert create_slogan('S@SH05') == 'S@SH05 drinks pepsi in his brand new BMW!'
```

Comment:
And again, please run the test cases...
TIP: return name from a wrapper (instead of True).

Reply:
Fixed it. I ran the test cases, went through the debugger - everything works correctly.